token-locker/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+use std::convert::TryInto;
+
+use admin_controlled::{AdminControlled, Mask};
+use near_contract_standards::fungible_token::metadata::FungibleTokenMetadata;
+use near_sdk::{AccountId, Balance, env, ext_contract, Gas, near_bindgen, PanicOnDefault, Promise, PromiseOrValue};
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::collections::UnorderedSet;
+use near_sdk::json_types::{U128, ValidAccountId};
+
+use bridge_common::{FT_TRANSFER_CALL_GAS, FT_TRANSFER_GAS, NO_DEPOSIT, parse_recipient, PAUSE_DEPOSIT, Recipient, ResultType, VERIFY_LOG_ENTRY_GAS};
+use bridge_common::prover::{EthAddress, ext_prover, Proof, validate_eth_address};
+
+use crate::unlock_event::EthUnlockedEvent;
+
+mod token_receiver;
+mod unlock_event;
+
+near_sdk::setup_alloc!();
+
+/// Gas to call finish withdraw method.
+/// This doesn't cover the gas required for calling transfer method.
+const FINISH_WITHDRAW_GAS: Gas = 30_000_000_000_000;
+
+/// Gas to call finish deposit method.
+const FT_FINISH_DEPOSIT_GAS: Gas = 10_000_000_000_000;
+
+/// Gas for fetching metadata of token.
+const FT_GET_METADATA_GAS: Gas = 10_000_000_000_000;
+
+/// Gas for emitting metadata info.
+const FT_FINISH_LOG_METADATA_GAS: Gas = 30_000_000_000_000;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    /// The account of the prover that we can use to prove.
+    pub prover_account: AccountId,
+    /// Ethereum address of the token bridge contract, in hex.
+    bridge_address: EthAddress,
+    /// Hashes of the events that were already used.
+    pub used_events: UnorderedSet<Vec<u8>>,
+    /// Mask determining all paused functions
+    paused: Mask,
```

Comment:
Let's rather incorporate the `Pausable` trait here defined in [near-plugins](https://github.com/aurora-is-near/near-plugins).
It would be good to use `Ownable` too.

Reply:
We need to update to near-sdk-rs v4 at first to use near-plugins.

Reply:
Created issues https://github.com/aurora-is-near/rainbow-token-connector/issues/125 and https://github.com/aurora-is-near/rainbow-token-connector/issues/126 to track this.